plugins/chat/app/serializers/chat/chatables_serializer.rb

```diff
@@ -0,0 +1,66 @@
+# frozen_string_literal: true
+
+module Chat
+  class ChatablesSerializer < ::ApplicationSerializer
+    attributes :users
+    attributes :direct_message_channels
+    attributes :category_channels
+
+    def users
+      (object.users || [])
+        .map do |user|
+          {
+            identifier: "u-#{user.id}",
+            model: ::Chat::ChatableUserSerializer.new(user, scope: scope, root: false),
+            type: "user",
+          }
+        end
+        .as_json
+    end
+
+    def direct_message_channels
+      user_ids = (object.users || []).map(&:id)
+      channels =
+        (object.direct_message_channels || []).reject do |channel|
+          channel_user_ids = channel.allowed_user_ids - [scope.user.id]
+          channel.allowed_user_ids.length == 1 &&
+            user_ids.include?(channel.allowed_user_ids.first) ||
+            channel_user_ids.length == 1 && user_ids.include?(channel_user_ids.first)
+        end
+
+      channels
+        .map do |channel|
+          {
+            identifier: "c-#{channel.id}",
+            type: "channel",
+            model:
+              ::Chat::ChannelSerializer.new(
+                channel,
+                scope: scope,
+                root: false,
+                membership: channel.membership_for(scope.user),
```

Comment:
These `membership_for` are a problem I need to solve, this is causing N+1s...

Reply:
So you would end up with extra steps in your service something like:

```ruby
step :fetch_memberships
step :serialize_category_channels
step :serialize_users
step :serialize_direct_message_channels

def fetch_memberships(guardian:, **)
   context.memberships = Chat::ChannelMembershipManager.all_for_user(guardian.user)
end

def serialize_category_channels(guardian:, memberships:, **)
  if context.category_channels.blank?
    context.category_channels = []
    return
  end

  context.category_channels = Chat::Action::SerializeChatables.call(chatables: context.category_channels, memberships: memberships, guardian: guardian, type: "channel")
end


# and so on...
```

Then the action could be like:

```ruby
def call(chatables:, guardian:, memberships:)
     chatables.map do |chatable|
        # You could just have different cases in each of these methods to return the correct thing
        {
          identifier: identifier(chatable),
          type: type(chatable),
          model: serialize(chatable, guardian, memberships),
        }
      end
      .as_json
end
```

The `serialize_direct_message_channels` could have the complex de-duplication code above.

Reply:
And if you don't like this being in the original service, in case you want the raw data, you could:

a) Make a separate `SerializeChatables` service to do all this
b) Update the original service to have `Query` classes for each step